pkg/local/diff.go

```diff
@@ -188,6 +186,59 @@ func Undo(c Changes) Changes {
 	return reversed
 }
 
+// WalkS3 - walk like an Egyptian... ¯\_(ツ)¯\_
+// This walker function simulates the way object listing is performed by S3. Contrary to how a standard FS walk function behaves, S3
+// does not take into consideration the directory hierarchy. Instead, object paths include the entire path relative to the root and as a result
+// the directory or "path separator" is also taken into account when providing the listing in a lexicographical order.
+func WalkS3(root string, callbackFunc func(p string, info fs.FileInfo, err error) error) error {
+	var dirQueue SortedQueue
+
+	err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if p == root {
+			return nil
+		}
+		if info.IsDir() {
+			// TODO: We don't return dir results for the listing, how will this effect directory markers, and can we even support directory markers?
+			// Save encountered directories in a sorted queue and compare them with the first appearance of an file in that level
+			heap.Push(&dirQueue, p+path.Separator) // add path separator to dir name and sort it later
+			return filepath.SkipDir
+		}
+
+		heap.Init(&dirQueue)
+		for dirQueue.Len() > 0 {
+			dir := heap.Pop(&dirQueue).(string)
+			if p > dir {
+				if err = WalkS3(dir, callbackFunc); err != nil {
+					return err
+				}
+			} else { // Put it back in its place
+				heap.Push(&dirQueue, dir)
+				break
+			}
```

Comment:
This adds a $\Omega(log n)$ factor to every iteration.  Why do we always remove the first heap element?  Just implement a "Peek" or "Top" method which returns element 0 of the heap array.